### PR TITLE
feat: add MCP server config ID to tool-call message parts

### PR DESCRIPTION
**Pull request by kylecarbs**
Adds an `mcp_server_config_id` field to `ChatMessagePart` for `tool-call` and `tool-result` parts. This lets the frontend identify which MCP server config a tool call originated from, enabling:

- Rendering the MCP server's custom icon next to its tool calls
- Humanizing tool names (e.g. `linear__list_issues` → "List issues")

**Backend changes:**

- `mcpToolWrapper` now carries its config ID and implements a new `MCPToolIdentifier` interface
- The chatloop wraps `publishMessagePart` to annotate streaming tool parts with the config ID
- The persistence layer in `chatd.go` annotates parts after `PartFromContent` conversion
- No database migration needed — the field is JSON-persisted in `chat_messages.content` and is optional/omitempty

**Frontend changes:**

- `mcpServerConfigId` propagates through the full parsing pipeline: `messageParsing` → `streamState` → `types` → `ConversationTimeline` → `Tool`
- `GenericToolRenderer` looks up the MCP server config by ID and passes `iconUrl` to `ToolIcon` and `mcpSlug` to `ToolLabel`
- `ToolIcon` renders an `ExternalImage` in a rounded container when an `iconUrl` is provided
- `ToolLabel` uses a new `humanizeMCPToolName` utility to strip the slug prefix and convert underscores to title-cased words
- MCP server configs are already fetched at the page level and threaded through to tool rendering

### CODE REVIEW

File: site/src/components/ai-elements/tool/Tool.tsx [hunk 1]
```diff
@@ -466,11 +475,26 @@ const GenericToolRenderer: FC<ToolRendererProps> = ({
 			}
 		: fileViewerOpts;
 
+	// Look up MCP server config for icon and slug.
+	const mcpServer =
+		mcpServerConfigId && mcpServers
```
Comment (mafredri) [inline site/src/components/ai-elements/tool/Tool.tsx]: Can't this use `mcpServers?.find`?

**Review by mafredri — APPROVED** [review]
Human review only caught a minor nit, second wave coming if you want to wait for it but other than that, approved.

File: site/src/components/ai-elements/tool/utils.ts [hunk 2]
```diff
@@ -555,6 +555,23 @@ export const buildEditDiff = (
 	return parsed[0].files[0];
 };
 
+/**
+ * Converts an MCP-prefixed tool name into a human-readable label.
+ * E.g. "linear__list_issues" with slug "linear" → "List issues"
+ */
+export function humanizeMCPToolName(
```
Comment (mafredri) [inline site/src/components/ai-elements/tool/utils.ts:562]: **P2** No tests for any new code path in this PR. *(Bisky P2, Hisoka P2, Mafu-san P2, Ryosuke P2)*

> `humanizeMCPToolName` is a new pure function with no tests, despite living next to a 700-line test file that covers every other exported utility. The function has observable edge cases: `humanizeMCPToolName('linear', 'linear__')` returns `""` (empty label), and `humanizeMCPToolName('linear', 'linear__get__data')` returns `"Get  data"` (double space).

> The `mcpServerConfigId` field is threaded through `parseMessageContent`, `mergeTools`, `applyMessagePartToStreamState`, and `buildStreamTools` with no test assertions in the existing test files for any of these functions.

> `MCPToolIdentifier` and `MCPServerConfigID()` are not tested in `mcpclient_test.go`. `TestConnectAll_DiscoverTools` checks prefixed names and descriptions but does not assert the returned tools implement the new interface.

The feature has zero regression safety net. `humanizeMCPToolName` is a pure function that's trivial to cover. The prop threading tests are one field addition per existing test case. The backend interface check is a single assertion in `TestConnectAll_DiscoverTools`.

File: coderd/x/chatd/chatloop/chatloop.go [hunk 3]
```diff
@@ -235,14 +237,32 @@ func Run(ctx context.Context, opts RunOptions) error {
...
 	}
 
+	tools := buildToolDefinitions(opts.Tools, opts.ActiveTools, opts.ProviderTools)
+
+	// Build a lookup from tool name to MCP server config ID
+	// so we can annotate published parts with the originating
+	// server.
+	toolNameToConfigID := make(map[string]uuid.UUID)
+	for _, t := range opts.Tools {
+		if mcp, ok := t.(mcpclient.MCPToolIdentifier); ok {
```
Comment (mafredri) [inline coderd/x/chatd/chatloop/chatloop.go]: **P2** New `chatloop → mcpclient` import breaks chatloop's previously provider-agnostic boundary. *(Ryosuke P2, Meruem P2, Robin P2, Hisoka P2)*

> Before this PR, chatloop imports only `fantasy`, `chatprompt`, `chatretry`, and `codersdk`. It works with `fantasy.AgentTool` without knowing tool provenance. The annotation in chatloop's `publishMessagePart` wrapper could instead live in chatd.go's `PublishMessagePart` callback, which chatd.go already controls. That consolidates the map to one site, removes chatloop's mcpclient import, and keeps the loop layer generic.

> *Meruem adds:* The identical pattern (build map, check `part.ToolName`, set `MCPServerConfigID`) appears in three sites total. If a fourth code path produces tool parts, it must independently remember to annotate, and nothing enforces this.

chatd.go already builds the same `toolNameToConfigID` map for the persistence path. Could the streaming annotation be done in chatd.go's callback that wraps `publishMessagePart`? That would remove the chatloop→mcpclient dependency and keep the map in one place.

File: site/src/components/ai-elements/tool/ToolIcon.tsx [hunk 4]
```diff
@@ -11,12 +12,28 @@ import {
...
+export const ToolIcon: React.FC<{
+	name: string;
+	isError: boolean;
+	iconUrl?: string;
+}> = ({ name, isError, iconUrl }) => {
 	const color = isError ? "text-content-destructive" : "text-content-secondary";
 	const base = cn("h-4 w-4 shrink-0", color);
+
+	// If an MCP icon URL is provided, render it.
+	if (iconUrl) {
```
Comment (mafredri) [inline site/src/components/ai-elements/tool/ToolIcon.tsx]: **P2** MCP tool icons lose the error state visual indicator. *(Mafuuu P2, Nami P2)*

> The component computes `color` from `isError` (line 20) but the early-return branch for `iconUrl` never applies it. Built-in tool icons render with `text-content-destructive` on error; MCP tool icons always render in a neutral `bg-surface-secondary` container regardless of error state.

> *Nami adds:* A tool call that errors out from an MCP server will show its cheerful custom icon with no visual signal that something went wrong.

The `color` and `base` variables on lines 20-21 are dead code in this branch. Adding a conditional red ring or border to the wrapper `div` when `isError` is true would restore parity with built-in icons.

File: site/src/components/ai-elements/tool/ToolIcon.tsx [hunk 5]
```diff
@@ -11,12 +12,28 @@ import {
...
+		return (
+			<div
+				className={cn(
+					"flex h-4 w-4 shrink-0 items-center justify-center",
+					"rounded-full bg-surface-secondary",
+				)}
+			>
+				<ExternalImage src={iconUrl} alt={`${name} icon`} className="h-3 w-3" />
```
Comment (mafredri) [inline site/src/components/ai-elements/tool/ToolIcon.tsx]: **P3** No fallback when `ExternalImage` fails to load. *(Nami P2, Hisoka P3)*

> `ExternalImage` is a thin wrapper around `` with no `onError` handler. If the MCP server's `icon_url` 404s or is unreachable, the browser renders a broken image placeholder inside the 16x16 container. There is no fallback to the default `WrenchIcon`.

The icon URL is admin-configured, so broken URLs should be rare. An `onError` handler that falls back to `WrenchIcon` would be a small defensive improvement.

File: site/src/components/ai-elements/tool/utils.ts [hunk 6]
```diff
@@ -555,6 +555,23 @@ export const buildEditDiff = (
...
+	slug: string,
+	prefixedName: string,
+): string {
+	const prefix = `${slug}__`;
+	const raw = prefixedName.startsWith(prefix)
+		? prefixedName.slice(prefix.length)
+		: prefixedName;
+	// Replace underscores with spaces, capitalize first letter.
+	const words = raw.replace(/_/g, " ").trim();
```
Comment (mafredri) [inline site/src/components/ai-elements/tool/utils.ts]: **P3** `humanizeMCPToolName` has edge cases that produce empty or malformed labels. *(Mafuuu P3, Luffy P2)*

> `humanizeMCPToolName("linear", "linear__")` strips the prefix, leaving `""`. `charAt(0).toUpperCase() + slice(1)` returns `""`. The label renders as a blank `<span>`.

> *Luffy adds:* A one-line guard would prevent invisible labels: `return words ? words.charAt(0).toUpperCase() + words.slice(1) : prefixedName`.

Also, `replace(/_/g, " ")` turns `__` into double spaces. Using `/_+/g` would collapse consecutive underscores. And camelCase names (`listIssues` becomes "Listissues") aren't handled, though that may be out of scope if MCP conventions always use underscores.

File: codersdk/chats.go [hunk 7]
```diff
@@ -146,25 +146,26 @@ func AllChatMessagePartTypes() []ChatMessagePartType {
...
-	FileID      uuid.NullUUID       `json:"file_id,omitempty" format:"uuid" variants:"file?"`
-	FileName    string              `json:"file_name" variants:"file-reference"`
-	StartLine   int                 `json:"start_line" variants:"file-reference"`
-	EndLine     int                 `json:"end_line" variants:"file-reference"`
+	Type              ChatMessagePartType `json:"type"`
+	Text              string              `json:"text" variants:"text,reasoning"`
+	Signature         string              `json:"signature,omitempty"`
+	ToolCallID        string              `json:"tool_call_id,omitempty" variants:"tool-call?,tool-result?"`
+	ToolName          string              `json:"tool_name,omitempty" variants:"tool-call?,tool-result?"`
+	MCPServerConfigID uuid.NullUUID       `json:"mcp_server_config_id,omitempty" format:"uuid" variants:"tool-call?,tool-result?"`
```
Comment (mafredri) [inline codersdk/chats.go:154]: **Obs** `omitempty` on `uuid.NullUUID` does not suppress the field on the wire. *(Bisky, Hisoka, Mafuuu, Nami, Ging)*

> Go's `encoding/json` does not consider struct types empty for `omitempty` purposes. `json.Marshal` on a zero `uuid.NullUUID` produces `null`, not an omitted field. Every non-MCP `ChatMessagePart` will include `"mcp_server_config_id": null`.

This is consistent with the existing `FileID` field (same type, same tag), so it follows codebase convention. Ging notes that `omitzero` (Go 1.24+) would be the correct tag, but changing it is a broader effort affecting both `NullUUID` fields plus the `TestChatMessagePartVariantTags` test. Not for this PR.

**Review by mafredri — COMMENTED** [review]
🤖 Agent on behalf of mafredri

Clean feature. The `MCPToolIdentifier` interface is a nice opt-in pattern that avoids coupling `fantasy.AgentTool` to MCP-specific concerns, and the frontend degrades gracefully when no MCP config is found. Threading the config ID through both streaming and persistence ensures the annotation survives regardless of the code path.

3 P2s, 2 P3s, 1 observation across 6 inline comments. Two main themes: missing test coverage for all new code paths, and the `ToolIcon` MCP branch that drops error styling and has no broken-image fallback.

> Fourteen personas, six findings, zero migrations. The best kind of backend change is the one that fits in a JSON column. Bra jobbat!

**Review by mafredri — COMMENTED** [review]
🤖 Agent on behalf of mafredri

All six findings addressed. The chatloop dependency removal is particularly clean, moving the annotation into chatd.go's callback consolidates the map and keeps chatloop provider-agnostic. The ToolIcon fallback with `useState` + `onError` is the right pattern. Tests cover the key edge cases.

LGTM from the panel.

> Six findings in, six fixes out, zero chatloop imports. Godkänt!
